packages/theme/middleware/checkout.js

```diff
@@ -0,0 +1,26 @@
+const canEnterBilling = cart => Boolean(cart.address?.shipping) && cart.lineItems.length;
+
+const canEnterPayment = cart => Boolean(cart.address?.billing) && Boolean(cart.address?.shipping) && cart.lineItems.length;
+
+export default async ({ app, $vsf }) => {
+  const currentPath = app.context.route.fullPath.split('/checkout/')[1];
+
+  if (!currentPath) return;
+
+  const cart = await $vsf.$spree.api.getCart();
+
+  if (!cart) return;
+
+  switch (currentPath) {
+    case 'billing':
```

Comment:
We should also handle a case when the cart is empty and the customer enters the first step of the checkout

```
case 'shipping':
   if (cartEmpty(cart)) {
     redirect to `/`
   }
```